pkg/trace/writer/sender.go

```diff
@@ -155,14 +157,34 @@ func (s *queuableSender) Start() {
 // Run executes the queuableSender main logic synchronously.
 func (s *queuableSender) Run() {
 	defer close(s.exit)
+	sema := make(chan struct{}, s.conf.MaxConnections)
 
 	for {
 		select {
 		case payload := <-s.in:
-			if stats, err := s.sendOrQueue(payload); err != nil {
-				log.Debugf("Error while sending or queueing payload. err=%v", err)
-				s.notifyError(payload, err, stats)
+			if payload == nil {
+				continue
+			}
+			s.mu.RLock()
+			queuing := s.queuedPayloads.Len() > 0
+			s.mu.RUnlock()
+			if queuing {
+				if err := s.enqueue(payload); err != nil {
+					log.Debugf("Error while queueing payload: %v", err)
+					s.notifyError(payload, err)
+				}
+				continue
 			}
+			sema <- struct{}{}
+			s.wg.Add(1)
+			go func() {
+				defer s.wg.Done()
+				if err := s.trySend(payload); err != nil {
+					log.Debugf("Error while sending or queueing payload. err=%v", err)
+					s.notifyError(payload, err)
+				}
+				<-sema
+			}()
```

Comment:
Why did we go the semaphore+1 routine per send route instead of having a pool of senders reading from an input channel?

Both approaches sound equivalent in behaviour but the first (implemented here) sounds a lot more expensive than the second in terms of runtime (creating routines is relatively cheap but not free)

Reply:
Honestly this has been the style of implementation used that I've seen elsewhere. I don't have any numbers of research to back it up. My first impressions as to why I chose this would be:
* It's basically dynamic and you don't have to make assumptions about how many workers are needed based on the system load (since various customers have different loads).
* Cost is negligible.
* Unused workers create pressure on the scheduler.

Reply:
> * It's basically dynamic and you don't have to make assumptions about how many workers are needed based on the system load (since various customers have different loads).

Yeah, this seems better than having a pool of 200 goroutines sitting around mostly doing nothing.